### PR TITLE
HPX version compatibility in spack package

### PR DESCRIPTION
**Pull request by albestro**
I realized this just now that I started using our spack package for DLAF.

It may be a bit late for this, since we are close to HPX 1.6.0, iirc. I'm going to add a similar thing to #315.

### CODE REVIEW

File: spack/packages/dla-future/package.py [hunk 1]
```diff
@@ -26,7 +26,7 @@ class DlaFuture(CMakePackage, CudaPackage):
     depends_on("blaspp")
     depends_on("lapackpp")
     depends_on("hpx cxxstd=14 networking=none")
-    depends_on("hpx@1.5.0", when="~cuda")
+    depends_on("hpx@1.5.0:1.5.99", when="~cuda")
```
Comment (msimberg) [inline spack/packages/dla-future/package.py:29]: Wasn't `hpx@1.5` equivalent to `hpx@1.5.<any patch version>`? Not that we're going to have more than 100 patch versions for 1.5, or more than 1 for that matter :D

**Review by teonnik — APPROVED** [review]
1.5.1 instead of 1.5.99 seems better but I don't see an issue either way.

File: spack/packages/dla-future/package.py (hunk 1 above)
Comment (albestro) [reply spack/packages/dla-future/package.py:29]: yeah...you're right! I always struggle with this...thanks for pointing it out!

**Comment by albestro** [comment]
@haampie with the change proposed by @msimberg, which looks absolutely the right thing to do, I'm addressing some problems.

The problem happens during the concretization, I'm going to give it a try to clingo.

However, what @teonnik suggested may be a nice solution, i.e. `@1.5.0:1.5.1`.

**Comment by albestro** [comment]
> @haampie with the change proposed by @msimberg, which looks absolutely the right thing to do, I'm addressing some problems.
> 
> The problem happens during the concretization, I'm going to give it a try to clingo.

Concretization with clingo works as expected with the versioning mechanism!
+1 for clingo! @haampie

**Comment by haampie** [comment]
I bumped this PR https://github.com/spack/spack/pull/20258, it's weird to not have closed-open version ranges

File: spack/packages/dla-future/package.py [hunk 2]
```diff
@@ -26,7 +26,10 @@ class DlaFuture(CMakePackage, CudaPackage):
     depends_on("blaspp")
     depends_on("lapackpp")
     depends_on("hpx cxxstd=14 networking=none")
-    depends_on("hpx@1.5.0", when="~cuda")
+    # it would have been "more correct" specifying @1.5 as constraint but, if clingo
+    # solver works as expected, the default solver has problems with that.
+    # Eventually, we opted for specifying the range more explicitly, so it works with both
+    depends_on("hpx@1.5.0:1.5.1", when="~cuda")
```
Comment (haampie) [inline spack/packages/dla-future/package.py:32]:
```suggestion
    depends_on("hpx@1.5.0:1.5.999", when="~cuda")
```
Just my 2 cents: it's horrible, but assuming hpx respects semver any future patch version should be ok, and whenever the open ranges PR lands in spack, it'll be easier to grep for `depends_on.*99` and fix it

**Pull request merged**